### PR TITLE
Update source build to sdk branch 20.08

### PR DESCRIPTION
**Pull request by jakobjakobson13**

### CODE REVIEW

**Comment by A6GibKm** [comment]
Seems you removed p7zip. I don't know if it is actually needed but if it is, the patch to make it compile with gcc10 is at: https://github.com/flathub/org.gnome.FileRoller/pull/2.

**Comment by jakobjakobson13** [comment]
> Seems you removed p7zip. I don't know if it is actually needed but if it is, the patch to make it compile with gcc10 is at: [flathub/org.gnome.FileRoller#2](https://github.com/flathub/org.gnome.FileRoller/pull/2).

I don't need it in this pull request but in the pull request #32 . So thank you very much.

**Comment by fedelibre** [comment]
@jakobjakobson13 I see that the aarch64 build is failing, but the details do not provide any information (stdio file seems empty...).

I may try building it on my machine.
Before I try, let me know if you're still working on this.
Thanks

**Comment by jakobjakobson13** [comment]
> @jakobjakobson13 I see that the aarch64 build is failing, but the details do not provide any information (stdio file seems empty...).
> 
> I may try building it on my machine.
> Before I try, let me know if you're still working on this.
> Thanks

Yes, the build is too old, so the log file got deleted. The easiest way to regain that log file is a forced rebuild of the manifest on the flathub builder. So, no need to waste your time and energy on that problem.

On the other hand, I was thinking about closing this repository, as tex editors can just point to a local texlive installation that gives the user more control and does not bloat each editor with a full texlive installation.

**Comment by fedelibre** [comment]
> On the other hand, I was thinking about closing this repository, as tex editors can just point to a local texlive installation that gives the user more control and does not bloat each editor with a full texlive installation.

Given the size of Texlive, it seems reasonable.

I'm using this extension to [build lilypond for Frescobaldi lilypond editor](https://github.com/flathub/org.frescobaldi.Frescobaldi/blob/master/org.frescobaldi.Frescobaldi.yaml#L10). If you close this repository I'm in trouble :-)

**Comment by jakobjakobson13** [comment]
The last lines of the `aarch64` log say
```
checking for pkg-config... pkg-config
checking for deflate... yes
checking zlib.h usability... yes
checking zlib.h presence... yes
checking for zlib.h... yes
checking for png_create_read_struct... no
configure: error: libpng not found
=== configuring in gd failed
make[2]: *** [Makefile:906: recurse] Error 1
make[2]: Leaving directory '/run/build/texlive-source/_flatpak_build/libs'
make[1]: *** [Makefile:483: all-recursive] Error 1
make[1]: Leaving directory '/run/build/texlive-source/_flatpak_build/libs'
make: *** [Makefile:577: all-recursive] Error 1
```

But be aware that this build contains only the very bare minimum of a texlive installation. (At least in my understanding.) So, it would take very much work to convert this manifest into a functioning texlive installation and I am not sure if it is worth the effort due to problematic bundling of a full texlive installation to each single tex editor.
In addition the folks at texlive did not seem to eager to put time into a flatpak package: [https://tug.org/pipermail/tex-live/2020-October/046290.html](https://tug.org/pipermail/tex-live/2020-October/046290.html), [https://tug.org/pipermail/tex-live/2020-October/046294.html](https://tug.org/pipermail/tex-live/2020-October/046294.html).

**Comment by hfiguiere** [comment]
It fails because:
```
configure:5963: checking for png_create_read_struct
configure:5963: gcc -o conftest -O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-
tables -fstack-clash-protection  -I/run/build/texlive-source/_flatpak_build/libs/libpng/include -I/run/build/texlive-source/_flatpak_build/libs/zlib/include  -L/app/lib -Wl,-z,r
elro,-z,now -Wl,--as-needed  conftest.c /run/build/texlive-source/_flatpak_build/libs/libpng/libpng.a /run/build/texlive-source/_flatpak_build/libs/zlib/libz.a -lm  >&5
/usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/../../../../aarch64-unknown-linux-gnu/bin/ld: /run/build/texlive-source/_flatpak_build/libs/libpng/libpng.a(pngrtran.o): in functio
n `png_do_expand_palette':
/run/build/texlive-source/_flatpak_build/libs/libpng/../../../libs/libpng/libpng-src/pngrtran.c:4323: undefined reference to `png_do_expand_palette_rgba8_neon'
/usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/../../../../aarch64-unknown-linux-gnu/bin/ld: /run/build/texlive-source/_flatpak_build/libs/libpng/libpng.a(pngrtran.o): in functio
n `png_do_read_transformations':
/run/build/texlive-source/_flatpak_build/libs/libpng/../../../libs/libpng/libpng-src/pngrtran.c:4782: undefined reference to `png_riffle_palette_neon'
/usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/../../../../aarch64-unknown-linux-gnu/bin/ld: /run/build/texlive-source/_flatpak_build/libs/libpng/libpng.a(pngrtran.o): in functio
n `png_do_expand_palette':
/run/build/texlive-source/_flatpak_build/libs/libpng/../../../libs/libpng/libpng-src/pngrtran.c:4354: undefined reference to `png_do_expand_palette_rgb8_neon'
/usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/../../../../aarch64-unknown-linux-gnu/bin/ld: /run/build/texlive-source/_flatpak_build/libs/libpng/libpng.a(pngrutil.o): in functio
n `png_init_filter_functions':
/run/build/texlive-source/_flatpak_build/libs/libpng/../../../libs/libpng/libpng-src/pngrutil.c:4126: undefined reference to `png_init_filter_functions_neon'
collect2: error: ld returned 1 exit status
configure:5963: $? = 1
```

A classic failure when building libpng from source. I'd vote for using the system libpng.

**Comment by hfiguiere** [comment]
I have submitted this https://github.com/jakobjakobson13/org.freedesktop.Sdk.Extension.texlive/pull/1 to allow this brnch to build on aarch64.

But with this, `xelatex` is missing. It is needed for https://github.com/flathub/org.cvfosammmm.Setzer

**Comment by hfiguiere** [comment]
using system libpng require -ldl (also for x86_64)
it also require to disable "native" for whatever it means, this changes the install path.
I'll follow up.

**Comment by jakobjakobson13** [comment]
If we include system libraries, we could also include much libraries more than only the `libpng` library. I'll post the list of libraries that are used in texlive and are provided by the freedesktop later.

**Comment by hfiguiere** [comment]
I was trying to minimise the difference. BTW `--disable-native-texlive-build` doesn't change the installation path like I said. With that build, the binaries end up in `/app/texlive/bin`, it's the install script in this PR. 

I'll update my branch.

**Comment by jakobjakobson13** [comment]
Xelatex only requires `fontconfig` for its build

```
xetex

    requires fontconfig (again both headers and library), or, for MacOSX only, the ApplicationServices and Cocoa frameworks.
```

So, I guess this would be the easiest part. In the following, however, we have to decide on whether we want to provide
- a bare/bone texlive extension providing only the binaries (follow-up question: how this can be extended individually by each single user that it matches her/his needs properly)
- a full texlive extension that provides ALL packages and documentation (that would increase the extension size to several GB of data which is probably an overkill for most uses)

An alternative "solution" would be that we archive the texlive extension with the recommendation to all maintainers allowing the use of a local texlive installation that has been performed with the `install-tl-unx` package.

**Comment by hfiguiere** [comment]
xetex did build without problem here.

the sdk is already around 3GB here on `x86_64`

**Comment by jakobjakobson13** [comment]
> the sdk is already around 3GB here on `x86_64`

I know. But as these 3 GB are currently shipped with **every** tex editor, we could try to find a smarter solution.

**Comment by hfiguiere** [comment]
I know they are shipped that way.

Two things however
- we could have it as a runtime extension
- flatpak does deduplicate, so in practice it's not using that much 

My goal here is just to have `aarch64` support.

**Comment by jakobjakobson13** [comment]
The build process worked well but know I get the following errors:
```
$ files/bin/x86_64/pdftex: symbol lookup error: ./pdftex: undefined symbol: kpathsea_cnf_line_env_progname
```
```
$ files/bin/x86_64/xetex: error while loading shared libraries: libicuuc.so.67: cannot open shared object file: No such file or directory
```
```
$ files/bin/x86_64/luatex: symbol lookup error: ./luatex: undefined symbol: kpathsea_cnf_line_env_progname
```

**Comment by hfiguiere** [comment]
I don't have these errors with my build using 

https://github.com/jakobjakobson13/org.freedesktop.Sdk.Extension.texlive/pull/2

**Comment by hfiguiere** [comment]
The problem is that the variable substitution isn't done. You need something like:

```yaml
built-options:
  arch:
    x86_64:
      prepend-path: /usr/lib/sdk/texlive/bin/x86_64-linux
    aarch64:
      prepend-path: /usr/lib/sdk/texlive/bin/aarch64-linux
```

I use prepend-path here because I believe it is preferable.

**Comment by hfiguiere** [comment]
Stream crossed, it was for the previous version.

**Comment by hfiguiere** [comment]
The current failure is about `Can't locate mktexlsr.pl ` is what I have been experiencing with running `xelatex` (if you pass a .tex file it fails that way).

**Comment by jakobjakobson13** [comment]
Yes, `mktexlsr.pl` is provided by `texlive-20200406-texmf.tar.xz`. And `texdoc` apparently needs the files from `texlive-20200406-tlpdb-full.tar.gz`.

**Comment by hfiguiere** [comment]
That's what I was suspecting
This page has the links for the two archives. http://www.linuxfromscratch.org/blfs/view/svn/pst/texlive.html

**Comment by hfiguiere** [comment]
With this PR the SDK build and install https://github.com/jakobjakobson13/org.freedesktop.Sdk.Extension.texlive/pull/3

**Comment by jakobjakobson13** [comment]
Thanks you very much for your help, @hfiguiere.

Now, we only have to take care of `asymptote`, `biber` and `dvisgm`. I'll try to work on it, once I found the reason why I can't install and test the latest build on my computer.

**Comment by hfiguiere** [comment]
With that packaging, I can tell that https://github.com/flathub/com.github.xournalpp.xournalpp/ will break since they use the SDK as a runtime.

**Comment by hfiguiere** [comment]
Locally I'm just missing Biber as I'm sort of blocked by https://github.com/flatpak/flatpak-builder-tools/issues/174 that I need to use to add all the Perl module dependencies.

**Comment by jakobjakobson13** [comment]
@hfiguiere Does xelatex work on your local test build? I still (no matter if I used the shared lib or not) get the previous error
```
$ files/bin/x86_64/xelatex: error while loading shared libraries: libicuuc.so.67: cannot open shared object file: No such file or directory
```

**Comment by hfiguiere** [comment]
Are you testing outside of the sandbox?

**Comment by hfiguiere** [comment]
Inside the sandbox, yes.

**Comment by jakobjakobson13** [comment]
My fault, no I did it outside the sandbox.

**Comment by hfiguiere** [comment]
It tested xelatex with org.cvfosammmm.Setzer that I rebuild with the newer SDK to test.

**Comment by hfiguiere** [comment]
~~The `x86_64` seems to be caused by the size of the SDK (I found it slows things down)~~
The `aarch64` failure, I will investigate. FWIW my PR are done from `aarch64`, it they are supposed to build.

**Comment by hfiguiere** [comment]
The x86_64 failure is that `~~xindy~~clisp expect libsigsegv, and it's not installed.

**Comment by hfiguiere** [comment]
Updated my MR, it should build now.

https://github.com/jakobjakobson13/org.freedesktop.Sdk.Extension.texlive/pull/5

**Comment by hfiguiere** [comment]
The build is working here, not sure what's happening *yet*

**Comment by jakobjakobson13** [comment]
At my PC currently two modules fail:
- dvisgm fails due to some missing woff libraries
- clisp fails due to the missing libsegv

The first failure could be solved by disabling woff, whereas I have no idea how to solve the second one.

**Comment by hfiguiere** [comment]
> At my PC currently two modules fail:
> 
>     * dvisgm fails due to some missing woff libraries
> 
>     * clisp fails due to the missing libsegv

This was part of my pull request. Are you sure you updated your tree?

The build went to the end on my RaspberryPi (I can't try on my laptop, I'm out of space right now)

**Comment by hfiguiere** [comment]
I see the woff error on the builder but not locally, not sure what's happening.

**Comment by jakobjakobson13** [comment]
Build 34681 is the first build to fail on x86_64 and the corresponding commit is `"add clisp and xindy (6e6f7076)"` that got fixed by you. The last build (34771, commit `"Fix clisp build (90c6cfd5)"`) does however fail due to dvisvgm, and this is odd as the dvisvgm module was not changed by the commits. So to me, this does not make much sense.

**Comment by jakobjakobson13** [comment]
The aarch64 build works now in contrast to the x86_64 build. The upload however fails due to a time out.

**Comment by hfiguiere** [comment]
https://github.com/jakobjakobson13/org.freedesktop.Sdk.Extension.texlive/pull/6 should fix the x86_64 build.

File: org.freedesktop.Sdk.Extension.texlive.yml [hunk 1]
```diff
@@ -61,60 +61,60 @@ modules:
...
-        md5: e1fa5d5ce656fedcdb7cd4146220e181
+        url: https://github.com/mgieseki/dvisvgm/archive/2.11.1.tar.gz
+        sha256: 71b772d97abe04bc0b9a4dbd0077f4224ade7ea0aac325f061e4c0037d31020c
   - name: clisp
     build-options:
       env:
-        CPPFLAGS: -I/usr/lib/sdk/texlive/include
-        LDFLAGS: -L/usr/lib/sdk/texlive/lib
+        cppflags: -I/usr/lib/sdk/texlive/include
+        ldflags: -L/usr/lib/sdk/texlive/lib
```
Comment (hfiguiere) [inline org.freedesktop.Sdk.Extension.texlive.yml]: These are environment variables. Hence the capitalization.

Comment (jakobjakobson13) [reply org.freedesktop.Sdk.Extension.texlive.yml]: Sorry, I was not aware of that.

**Comment by fedelibre** [comment]
Kudos to @jakobjakobson13 and @hfiguiere for this epic PR!

**Comment by jakobjakobson13** [comment]
As I said, the thank goes mostly to @hfiguiere

**Pull request merged**